### PR TITLE
Revamp build history search bar

### PR DESCRIPTION
**Pull request by janfaracik**
### Proposed changelog entries

* Modernise the "Build History" search bar

### Video

https://user-images.githubusercontent.com/43062514/130856620-48c47dc0-4385-4313-8194-050ec0ad07ba.mov

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by Wadeck — COMMENTED** [review]
Not manually tested but looks really nice :)

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (fqueiruga) [inline war/src/main/webapp/scripts/hudson-behavior.js:2294]: Cannot use `const` on the `hudson-behavior.js` file as it's not processed through webpack. These should be `var`s

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (fqueiruga) [inline war/src/main/webapp/scripts/hudson-behavior.js:2338]: I'd suggest to revert this. This would cause a lot of requests to the backend, and there's no sane way of adding a debounce to this file. Other option would be to extract all this functionality to a webpack entrypoint and use `lodash.debounce`

**Review by fqueiruga — CHANGES_REQUESTED** [review]

**Comment by fqueiruga** [comment]
Looks good. I do have some comments about some issues that will arise.

I'd also suggest using, if possible, the `.setting-input` CSS class to the input to provide the basic styles, and only add the needed styles on top. That way it will be updated if forms input ever get revamped.

**Comment by oleg-nenashev** [comment]
@janfaracik Looks like this one is almost ready to go. WDYT about the comments by @fqueiruga ?

**Comment by janfaracik** [comment]
> @janfaracik Looks like this one is almost ready to go. WDYT about the comments by @fqueiruga ?

They sound sensible to me 👍 I'll be making some changes, I'll post another comment when it should be good for another look over.

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (janfaracik) [reply war/src/main/webapp/scripts/hudson-behavior.js]: Done.

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (janfaracik) [reply war/src/main/webapp/scripts/hudson-behavior.js]: I've extracted the build history JS from `hudson-behavior.js` and added debounce - the file itself is a little rough around the edges at the moment but its working well from the look of it.

**Comment by janfaracik** [comment]
I've updated the PR based on the reviews - thanks everybody by the way :) 

I've added debounce so server calls are limited, there's a short delay after typing before calling the API so it shouldn't swamp the API. I've also added a small animation to indicate that it's loading and I've also a small notice when there are no builds.

https://user-images.githubusercontent.com/43062514/133934061-47a4e67d-1e67-4306-bcc7-40a6efb9e96b.mov

Reviews are welcome, I've extracted the JS into its own file now which makes things cleaner and easier to maintain - the file could definitely be cleaner though so would be great for people to try out the PR/suggest improvements.

Cheers

**Comment by timja** [comment]
Wow I just tried out the old behaviour and that is not a nice UX, never used this before tbh

**Comment by timja** [comment]
Icon appears to be mis-aligned for me?

Running chrome on mac big sur

image

**Comment by timja** [comment]
Spacing appears to be off for me too, it doesn't appear like in the video above.

The `x` button doesn't change the cursor to show it's clickable.

doesn't look the best in dark theme but looks like there's variables there that can be used to tweak that

**Comment by janfaracik** [comment]
> Spacing appears to be off for me too, it doesn't appear like in the video above.
> 
> The `x` button doesn't change the cursor to show it's clickable.
> 
> doesn't look the best in dark theme but looks like there's variables there that can be used to tweak that

Darn Safari handles padding/line height _**differently**_ for search inputs :doh: Should be all fixed now :) 

The clear button also now changes the cursor on hover.

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (fqueiruga) [reply war/src/main/webapp/scripts/hudson-behavior.js]: Ok, sounds good

**Review by timja — APPROVED** [review]
the focus appears to be quite strong (and different from the rest of the application). Not that I like the focus state in the rest of the application.

I think in this case it looks fine, although if it was rolled out more widely it would probably be too 'much' when there's a number of inputs together, but I could be wrong.

My previous feedback was addressed and this is a **lot** better than the previous search bar

**Comment by timja** [comment]
conflicts

**Comment by timja** [comment]
@fqueiruga any chance you could take another look?

**Review by fqueiruga — APPROVED** [review]
I think this is a great PR. I'm still a bit concious about the use of `px` but that's not a blocker.

Cheers @janfaracik , this is really nice.

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by dheerajodha** [comment]
Suggested Edit:

- Modernise the <code>Build History</code> search bar.

**Comment by timja** [comment]
Possibly caused https://issues.jenkins.io/browse/JENKINS-66753

**Comment by janfaracik** [comment]
> Possibly caused https://issues.jenkins.io/browse/JENKINS-66753

Investing this now...

**Pull request merged**
